File: 백준/Gold/1976. 여행 가자/여행 가자.py
n = int(input())
m = int(input())
arr = [list(map(int, input().split())) for _ in range(n)]
arrive = list(map(int, input().split()))
parent = [i for i in range(n)]
size =  [0] * n

def find(x):
    # 처음에 똑같은 루트 만나면 갱신
    if x != parent[x]:
        parent[x] = find(parent[x])
    return parent[x]


def unite(a, b):
    ra, rb = find(a), find(b)

    # 길이 크기 비교
    if size[rb] > size[ra]:
        ra, rb = rb, ra

    parent[rb] = parent[ra] # 부모 갱신
    size[ra] += size[rb] # size 갱신,

    return


for i in range(n):
    for j in range(n):
        if arr[i][j] == 1:
            unite(i, j)


diff = find(arrive[0] - 1)
ans = "YES"
for i in arrive:
    i -= 1  # 0 base 에 맞춰야 함
    if find(i) != diff:
        ans = "NO"
        break

print(ans)

File: 백준/Gold/1976. 여행 가자/여행 가자.cc
#include <iostream>
#include <bits/stdc++.h>
using namespace std;
vector<int> parent, sz;  // 전역 parent 배열

// find 함수 (경로 압축 포함)
int findUF(int x) {
    if (parent[x] != x) parent[x] = findUF(parent[x]);
    return parent[x];
}

// union 함수
void unite(int a, int b) {
    int ra = findUF(a),  rb = findUF(b);
    if (ra == rb) return;
    // size 확인 후 , ra, rb 갱신
    if (sz[rb] > sz[ra]) swap(ra, rb);

    // parent 갱신 size 갱신
    parent[rb] = parent[ra];
    sz[ra] += sz[rb];
}



int main() {
    // freopen("input.txt", "r", stdin);
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    // n, m
    int n, m;
    cin >> n >> m;
    // 인접 행렬
    vector<vector<int>> arr(n, vector<int>(n));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> arr[i][j];
        }
    }

    // arrive
    vector<int> arrive(m);
    for (int i = 0; i < m; i++) {
        cin >> arrive[i];
        --arrive[i]; // 0 base 처리
    }

    // resize
    parent.resize(n);
    sz.resize(n);
    // 0부터 n까지 parent 초기값 갱신
    iota(parent.begin(), parent.end(), 0);

    // union 실행
    for (int i = 0; i < n; i++) {
        for (int j =0; j < n; j++) {
            if (arr[i][j] == 1) {
                unite(i, j);
            }
        }
    }

    // find하기
    string ans = "YES";
    int target = findUF(arrive[0]);
    for (int i = 1; i < m; i++) {
        if (target != findUF(arrive[i])) {
            ans = "NO";
            break;
        }
    }
    cout << ans << endl;;
}

